README.md
Python WebHDFS
==============

WebHDFS python client library and simple shell.


Table of Contents
-----------------
* [Prerequisites](#prerequisites)
* [Installation](#installation)
* [API](#api)
  * [WebHDFSClient](#webhdfsclient)
    * [`__init__()`](#__init__base-user-confnone-waitnone)
    * [`stat()`](#statpath-catchfalse)
    * [`ls()`](#lspath-recursefalse-requestfalse)
    * [`glob()`](#globpath)
    * [`du()`](#dupath-realfalse)
    * [`mkdir()`](#mkdirpath)
    * [`mv()`](#mvpath-dest)
    * [`rm()`](#rmpath)
    * [`repl()`](#replpath-num)
    * [`get()`](#getpath-datanone)
    * [`put()`](#putpath-data)
    * [`content_summary()`](#get-content-summary)
  * [WebHDFSObject](#webhdfsobject)
    * [`__init__()`](#__init__path-bits)
    * [`is_dir()`](#is_dir)
    * [`is_empty()`](#is_empty)
    * [`owner`](#owner)
    * [`group`](#group)
    * [`name`](#name)
    * [`full`](#full)
    * [`size`](#size)
    * [`repl`](#repl)
    * [`kind`](#kind)
    * [`date`](#date)
    * [`mode`](#mode)
    * [`perm`](#perm)
* [Usage](#usage)
* [License](#license)


Prerequisites
-------------
* Python 2.7+
* Python [requests](http://docs.python-requests.org/) module


Installation
------------
Install python-webhdfs as a Debian package by building a deb:

    dpkg-buildpackage
    # or
    pdebuild

Install python-webhdfs using the standard setuptools script:

    python setup.py install


API
---
To use the WebHDFS Client API, start by importing the class from the module

```python
>>> from webhdfs import WebHDFSClient
```

All functions may throw a `WebHDFSError` exception or one of these subclasses:

| Exception Type                   | Remote Exception              | Description                                |
|----------------------------------|-------------------------------|--------------------------------------------|
| WebHDFSConnectionError           |                               | Unable to connect to active NameNode       |
| WebHDFSIncompleteTransferError   |                               | Transferred file doesn't match origin size |
| WebHDFSAccessControlError        | AccessControlException        | Access to specified path denied            |
| WebHDFSIllegalArgumentError      | IllegalArgumentException      | Invalid parameter value                    |
| WebHDFSFileNotFoundError         | FileNotFoundException         | Specified path does not exist              |
| WebHDFSSecurityError             | SecurityException             | Failed to obtain user/group information    |
| WebHDFSUnsupportedOperationError | UnsupportedOperationException | Requested operation is not implemented     |
| WebHDFSUnknownRemoteError        |                               | Remote exception unrecognized              |

## `WebHDFSClient` ##

#### `__init__(base, user, conf=None, wait=None)` ####
Creates a new `WebHDFSClient` object

Parameters:
* `base`: base webhdfs url. (e.g. http://localhost:50070)
* `user`: user name with which to access all resources
* `conf`: (_optional_) path to hadoop configuration directory for NameNode HA resolution
* `wait`: (_optional_) floating point number in seconds for request timeout waits

```python
>>> import getpass
>>> hdfs = WebHDFSClient('http://localhost:50070', getpass.getuser(), conf='/etc/hadoop/conf', wait=1.5)
```

#### `stat(path, catch=False)` ####
Retrieves metadata about the specified HDFS item.  Uses this WebHDFS REst request:

    GET <BASE>/webhdfs/v1/<PATH>?op=GETFILESTATUS

Parameters:
* `path`: HDFS path to fetch
* `catch`: (_optional_) trap `WebHDFSFileNotFoundError` instead of raising the exception

Returns:
* A single [`WebHDFSObject`](#webhdfsobject) object for the specified path.
* `False` if object not found in HDFS and `catch=True`.

```python
>>> o = hdfs.stat('/user')
>>> print o.full
/user
>>> print o.kind
DIRECTORY
>>> o = hdfs.stat('/foo', catch=True)
>>> print o
False
```


#### `ls(path, recurse=False, request=False)` ####
Lists a specified HDFS path.  Uses this WebHDFS REst request:

    GET <BASE>/webhdfs/v1/<PATH>?op=LISTSTATUS

Parameters:
* `path`: HDFS path to list
* `recurse`: (_optional_) descend down the directory tree
* `request`: (_optional_) filter request callback for each returned object

Returns:
* Generator producing children [`WebHDFSObject`](#webhdfsobject) objects for the specified path.

```python
>>> l = list(hdfs.ls('/')) # must convert to list if referencing by index
>>> print l[0].full
/user
>>> print l[0].kind
DIRECTORY
>>> l = list(hdfs.ls('/user', request=lambda x: x.name.startswith('m')))
>>> print l[0].full
/user/max
```


#### `glob(path)` ####
Lists a specified HDFS path pattern.  Uses this WebHDFS REst request:

    GET <BASE>/webhdfs/v1/<PATH>?op=LISTSTATUS

Parameters:
* `path`: HDFS path pattern to list

Returns:
* List of [`WebHDFSObject`](#webhdfsobject) objects for the specified pattern.

```python
>>> l = hdfs.glob('/us*')
>>> print l[0].full
/user
>>> print l[0].kind
DIRECTORY
```


#### `du(path, real=False)` ####
Gets the usage of a specified HDFS path.  Uses this WebHDFS REst request:

    GET <BASE>/webhdfs/v1/<PATH>?op=GETCONTENTSUMMARY

Parameters:
* `path`: HDFS path to analyze
* `real`: (_optional_) return machine usage instead of HDFS usage, taking replication factor into account

Returns:
* Integer of bytes used by the specified path.

```python
>>> u = hdfs.du('/user')
>>> print u
110433
```


#### `mkdir(path)` ####
Creates the specified HDFS path.  Uses this WebHDFS rest request:

    PUT <BASE>/webhdfs/v1/<PATH>?op=MKDIRS

Parameters:
* `path`: HDFS path to create

Returns:
* Boolean `True`

```python
>>> hdfs.mkdir('/user/%s/test' % getpass.getuser())
True
```


#### `mv(path, dest)` ####
Moves/renames the specified HDFS path to specified destination.  Uses this WebHDFS rest request:

    PUT <BASE>/webhdfs/v1/<PATH>?op=RENAME&destination=<DEST>

Parameters:
* `path`: HDFS path to move/rename
* `dest`: Destination path

Returns:
* Boolean `True` on success and `False` on error

```python
>>> hdfs.mv('/user/%s/test' % getpass.getuser(), '/user/%s/test.old' % getpass.getuser())
True
>>> hdfs.mv('/user/%s/test.old' % getpass.getuser(), '/some/non-existant/path')
False
```


#### `rm(path)` ####
Removes the specified HDFS path.  Uses this WebHDFS rest request:

    DELETE <BASE>/webhdfs/v1/<PATH>?op=DELETE

Parameters:
* `path`: HDFS path to remove

Returns:
* Boolean `True`

```python
>>> hdfs.rm('/user/%s/test' % getpass.getuser())
True
```


#### `repl(path, num)` ####
Sets the replication factor for the specified HDFS path.  Uses this WebHDFS rest request:

    PUT <BASE>/webhdfs/v1/<PATH>?op=SETREPLICATION

Parameters:
* `path`: HDFS path to change
* `num`:  new replication factor to apply

Returns:
* Boolean `True` on success, `False` otherwise

```python
>>> hdfs.stat('/user/%s/test' % getpass.getuser()).repl
1
>>> hdfs.repl('/user/%s/test' % getpass.getuser(), 3).repl
True
>>> hdfs.stat('/user/%s/test' % getpass.getuser()).repl
3
```


#### `get(path, data=None)` ####
Fetches the specified HDFS path.  Returns a string or writes a file, based on parameters.  Uses this WebHDFS request:

    GET <BASE>/webhdfs/v1/<PATH>?op=OPEN

Parameters:
* `path`: HDFS path to fetch
* `data`: (_optional_) file-like object open for write

Returns:
* Boolean `True` if data is set and written file size matches source
* String contents of the fetched file if data is None

Raises:
* `WebHDFSIncompleteTransferError`


#### `put(path, data)` ####
Creates the specified HDFS file using the contents of a file open for read, or value of the string.  Uses this WebHDFS request:

    PUT <BASE>/webhdfs/v1/<PATH>?op=CREATE

Parameters:
* `path`: HDFS path to fetch
* `data`: file-like object open for read or string

Returns:
* Boolean `True` if written file size matches source

Raises:
* `WebHDFSIncompleteTransferError`

#### `content_summary(path)` ####
Get Content Summary of a Directory.  Returns a dict.  Uses this WebHDFS request:

    GET <BASE>/webhdfs/v1/<PATH>?op=GETCONTENTSUMMARY

Parameters:
* `path`: HDFS directory to get the content summary for

Returns:
* Dict containg the Content Summary of the specified path or None on error

```python
>>> hdfs.content_summary('/user/test')
{
    "directoryCount": 2,
    "fileCount"     : 1,
    "length"        : 24930,
    "quota"         : -1,
    "spaceConsumed" : 24930,
    "spaceQuota"    : -1
}
```

## `WebHDFSObject` ##

#### `__init__(path, bits)` ####
Creates a new `WebHDFSObject` object

Parameters:
* `path`: HDFS path prefix
* `bits`: dictionary as returned by [`stat()`](#statpath-catchfalse) or [`ls()`](#lspath-recursefalse-requestfalse) call.

```python
>>> o = hdfs.stat('/')
>>> type(o)
<class 'webhdfs.attrib.WebHDFSObject'>
```

#### `is_dir()` ####
Determines whether the HDFS object is a directory or not.

Parameters: None

Returns:
* boolean `True` when object is a directory, `False` otherwise

```python
>>> o = hdfs.stat('/')
>>> o.is_dir()
True
```

#### `is_empty()` ####
Determines whether the HDFS object is empty or not.

Parameters: None

Returns:
* boolean `True` when object is a directory and has no children or a file and is of 0 size, and `False` otherwise

```python
>>> o = hdfs.stat('/')
>>> o.is_empty()
False
```


#### `owner` ####
Read-only property that retreives the HDFS object owner.

```python
>>> o = hdfs.stat('/')
>>> o.owner
u'hdfs'
```

#### `group` ####
Read-only property that retreives the HDFS object group.

```python
>>> o = hdfs.stat('/')
>>> o.group
u'supergroup'
```

#### `name` ####
Read-only property that retreives the HDFS object base file name.

```python
>>> o = hdfs.stat('/user/max')
>>> o.name
'max'
```

#### `full` ####
Read-only property that retreives the HDFS object full file name.

```python
>>> o = hdfs.stat('/user/max')
>>> o.full
'/user/max'
```

#### `size` ####
Read-only property that retreives the HDFS object size in bytes.

```python
>>> o = hdfs.stat('/user/max/snmpy.mib')
>>> o.size
20552
```

#### `repl` ####
Read-only property that retreives the HDFS object replication factor.

```python
>>> o = hdfs.stat('/user/max/snmpy.mib')
>>> o.repl
1
```

#### `kind` ####
Read-only property that retreives the HDFS object type (`FILE` or `DIRECTORY`).

```python
>>> o = hdfs.stat('/user/max/snmpy.mib')
>>> o.kind
u'FILE'
```

#### `date` ####
Read-only property that retreives the HDFS object last modification timestamp as a Python [datetime](https://docs.python.org/2/library/datetime.html) object.

```python
>>> o = hdfs.stat('/user/max/snmpy.mib')
>>> o.date
datetime.datetime(2015, 3, 7, 3, 53, 6)
```

#### `mode` ####
Read-only property that retreives the HDFS object symbolic permissions mode.

```python
>>> o = hdfs.stat('/user/max/snmpy.mib')
>>> o.mode
'-rw-r--r--'
```

#### `perm` ####
Read-only property that retreives the HDFS object octal permissions mode, usable by Python's [stat](https://docs.python.org/2/library/stat.html#module-stat) module.

```python
>>> o = hdfs.stat('/user/max/snmpy.mib')
>>> oct(o.perm)
'0100644'
>>> stat.S_ISDIR(o.perm)
False
>>> stat.S_ISREG(o.perm)
True
```

Usage
-----
```
usage: webhdfs [-h] [-d CWD] [-l LOG] [-c CFG] [-t TIMEOUT] [-v]
               url [cmd [cmd ...]]

webhdfs shell

positional arguments:
  url                   webhdfs base url
  cmd                   run this command and exit

optional arguments:
  -h, --help            show this help message and exit
  -d CWD, --cwd CWD     initial hdfs directory
  -l LOG, --log LOG     logger destination url
  -c CFG, --cfg CFG     hdfs configuration dir
  -t TIMEOUT, --timeout TIMEOUT
                        request timeout in seconds
  -v, --version         print version and exit

supported logger formats:
  console://?level=LEVEL
  file://PATH?level=LEVEL
  syslog+tcp://HOST:PORT/?facility=FACILITY&level=LEVEL
  syslog+udp://HOST:PORT/?facility=FACILITY&level=LEVEL
  syslog+unix://PATH?facility=FACILITY&level=LEVEL
```

Parameters:
* `url`: base url for the WebHDFS endpoint, supporting http, https, and hdfs schemes
* `cmd`: (_optional_) run the specified command with args and exit without starting the shell
* `-d | --cwd`: (_optional_) initial hdfs directory to switch to on shell invocation
* `-l | --log`: (_optional_) logger destination url as described by supported formats
* `-c | --cfg`: (_optional_) hadoop configuration directory for NameNode HA resolution
* `-t | --timeout`: (_optional_) request timeout in seconds as floating point number
* `-v | --version`: (_optional_) print shell/library version and exit

Environment Variables:
* `HADOOP_CONF_DIR`: alternative to and takes precedence over the `-c | --cfg` command-line parameter

License
-------
[MIT](http://mk23.mit-license.org/2015-2016/license.html)


debian/changelog
webhdfs (1.0.8) stable; urgency=low

  * Tagging 1.0.8
  * [adc375e] Issue #32: Add GETCONTENTSUMMARY support (Aaron Turner)

 -- Aaron Turner <synfinatic@gmail.com>  Mon, 08 Aug 2016 15:45:00 -0800

webhdfs (1.0.7) stable; urgency=medium

  * Tagging 1.0.7
  * [e1f5794] Issue #28: Update rmdir shell command to use is_empty()
    since ls() now always returns a generator which is a boolean True.
    (Max Kalika)
  * [2f19a76] Issue #27: Document is_empty() attribute API call. (Max
    Kalika)
  * [fee918c] Issue #27: Add is_empty() attribute call for files and
    directories. (Max Kalika)
  * [8f51620] Issue #28: Document ls() API call change to generator.
    (Max Kalika)
  * [a99da8e] Issue #28: convert ls() API call to a generator. (Max
    Kalika)
  * [66c8175] Issue #26: Update license year. (Max Kalika)
  * [86bbfb7] Issue #19: Add shell mv command. (Max Kalika)
  * [22938e7] Issue #24: Rewrite completion to use help docs usage
    patterns to automatically determine source and limits. (Max Kalika)
  * [2397ffc] Issue #23: Add help documentation to shell commands. (Max
    Kalika)
  * [c99a906] Issue #22: Simplify object existence tests in shell by
    using new stat() API catch parameter. (Max Kalika)
  * [68138e5] Issue #22: Document stat() API function catch parameter.
    (Max Kalika)
  * [b35620d] Issue #22: Add catch parameter to stat() API call to
    return False on WebHDFSFileNotFoundError instead of raising the
    exception. (Max Kalika)
  * [ff9897f] Issue #25: Document ls() API function request callback
    parameter. (Max Kalika)
  * [6e36fee] Issue #25: Add filtering callback to ls() API call. (Max
    Kalika)
  * [66f5eb1] Issue #21: Document put() API function. (Max Kalika)
  * [0534af1] Issue #20: Document new error types. (Max Kalika)
  * [3b7aba5] Issue #20: Add and use specific error types. (Max Kalika)
  * [895bc46] Issue #19: Document mv() API function. (Max Kalika)
  * [a2f2ee4] Issue #19: Add mv() API call to move/rename HDFS objects.
    (Max Kalika)
  * [fe2f916] Issue #18: Fix repl() documentation. (Max Kalika)

 -- Max Kalika <max.kalika+projects@gmail.com>  Thu, 07 Jan 2016 10:12:53 -0800

webhdfs (1.0.6) stable; urgency=medium

  * Tagging 1.0.6
  * [29befa2] Issue #17: Add glob command to the shell.
  * [b96b39d] Issue #16: Document glob() API function.
  * [5fcb0be] Issue #16: Add glob() API call to return HDFS objects
    matching requested pattern.

 -- Max Kalika <max.kalika+projects@gmail.com>  Fri, 09 Oct 2015 10:01:10 -0700

webhdfs (1.0.5) stable; urgency=medium

  * Tagging 1.0.5
  * [8806313] Issue #15: add shell lsr (recursive ls) command.
  * [8ecbdff] Issue #12: Use sys.stdout.write() instead of print for cat
    and zcat commands.
  * [0e88302] Issue #13: Document --version parameter.
  * [8449498] Issue #13: Add --version parameter to shell script.
  * [2ccb6bd] Issue #14: process cwd parameter before oneshot commands.

 -- Max Kalika <max.kalika+projects@gmail.com>  Wed, 26 Aug 2015 09:05:57 -0700

webhdfs (1.0.4) stable; urgency=medium

  * Tagging 1.0.4
  * [bfdfbeb] Issue #11: Sanitize paths in the client api.
  * [1a4c117] Issue #9: Document one-shot command support.
  * [15338ad] Issue #9: Add one-shot command support to the shell.
  * [192f895] Issue #8: Add parameter for initial HDFS directory to
    switch to on shell invocation.
  * [d91d5d7] Issue #10: Call raise_for_status() earlier in fetch
    requests to avoid masking http errors with other exceptions.

 -- Max Kalika <max.kalika+projects@gmail.com>  Fri, 31 Jul 2015 14:58:16 -0700

webhdfs (1.0.3) stable; urgency=medium

  * Tagging 1.0.3
  * [d173bbc] Issue #6: Document API and shell timeout feature.
  * [a799c64] Issue #6: Add wait parameter to WebHDFS shell.
  * [026584e] Issue #6: Add wait parameter to the WebHDFS client.
  * [19cb24d] Issue #7: Avoid crashing when given malformed xml to
    parse.
  * [8a036d4] Issue #5: Avoid crashing after failing to resolve
    namenodes with given configuration. Instead uuse the supplied URL as
    is.

 -- Max Kalika <max.kalika+projects@gmail.com>  Fri, 08 May 2015 09:23:35 -0700

webhdfs (1.0.2) stable; urgency=medium

  * Tagging 1.0.2
  * [8357aad] Issue #4: Add python-requests to debian build-depends to
    fix pbuilder.

 -- Max Kalika <max.kalika+projects@gmail.com>  Tue, 05 May 2015 17:44:31 -0700

webhdfs (1.0.1) stable; urgency=medium

  * Tagging 1.0.1
  * [d486575] Issue #2: Document shell usage.
  * [9e6e59e] Issue #3: Fix logger format output wrapping in shell's
    help.
  * [6952930] Issue #1: Document NameNode HA configuration support.
  * [981502a] Issue #1: Add support for specifying config path via
    HADOOP_CONF_DIR env var.
  * [05884db] Issue #1: Add --cfg flag in command pass to WebHDFSPrompt
    class.
  * [b6cc7e0] Issue #1: Failover to HA namenode on connection failures,
    timeouts, and standby exceptions.
  * [7ca9d18] Issue #1: Parse hadoop xml configs to extract HA
    namenodes.

 -- Max Kalika <max.kalika+projects@gmail.com>  Tue, 05 May 2015 15:33:35 -0700

webhdfs (1.0) stable; urgency=medium

  * Tagging initial 1.0
  * [571f7f9] Remove - from readline's completer delims, fixing objects
    with dashes in the name.
  * [902ea39] Add debian and release support.
  * [735c3eb] Add documentation and license.
  * [6ea361d] Move stat() closer to ls().
  * [91816bc] Fix string-based upload by rewinding the temp file handle
    before handing it to requests.
  * [3a0704e] Mark _log and _req methods as private.
  * [c8b855b] Detect directory vs file attributes.
  * [2d58b5a] Fix shell client library import.
  * [9debed5] Make other API classes available from base module.
  * [c203a27] Rename object.py -> attrib.py and clean up base imports.
  * [c1c3b42] Migrate LocalFSObject to object.py.  Fix ls, lls, cd, lcd
    shell commands.  Throw already exist errors in mkdir and put shell
    commands.
  * [76dfc0e] Remove pwd and lpwd shell commands since paths are
    displayed in the prompt.

lib/client.py
import errno
import fnmatch
import json
import logging
import os
import requests
import tempfile
import urlparse
import xml.etree.cElementTree as ET

from errors import WebHDFSError, WebHDFSConnectionError, WebHDFSFileNotFoundError, WebHDFSIncompleteTransferError
from attrib import WebHDFSObject

LOG = logging.getLogger()

class WebHDFSClient(object):
    def __init__(self, base, user, conf=None, wait=None):
        self.user = user
        self.wait = wait or 0.5
        self._cfg(base, conf)


    def _url(self, url):
        if url.scheme == 'hdfs':
            url = url._replace(scheme='http', netloc='%s:%s' % (url.hostname, url.port or '50070'))

        return url.geturl()

    def _cfg(self, base, conf=None):
        url = urlparse.urlparse(base)
        self.urls = []

        for part in ('hdfs', 'core'):
            item = '%s/%s-site.xml' % (os.environ.get('HADOOP_CONF_DIR', conf), part)
            LOG.debug('parsing %s for %s', item, url.hostname)

            try:

                tree = ET.parse(item).getroot()
                name = tree.find('.//property/[name="dfs.ha.namenodes.%s"]/value' % url.hostname)

                if name is None:
                    continue

                LOG.debug('found ha namenodes: %s', name.text)
                for host in name.text.split(','):
                    addr = tree.find('.//property/[name="dfs.namenode.http-address.%s.%s"]/value' % (url.hostname, host.strip()))
                    self.urls.append(self._url(url._replace(netloc=addr.text.strip())))
                    LOG.debug('resolved namenode address: %s -> %s', host.strip(), addr.text.strip())

                if self.urls:
                    break
            except ET.ParseError:
                LOG.debug('%s: failed to parse as xml', item)
                continue
            except EnvironmentError as e:
                if e.errno == errno.ENOENT:
                    LOG.debug('%s: file not found', item)
                    continue
                else:
                    raise e
        else:
            self.urls.append(self._url(url))

    def _req(self, name, path, kind='get', data=None, **args):
        args['op']        = name
        args['user.name'] = self.user

        try:
            for indx, base in enumerate(self.urls):
                u = '%s/webhdfs/v1/%s' % (base, requests.compat.quote(path.lstrip('/')))
                try:
                    if not data:
                        r = getattr(requests, kind)(u, params=args, timeout=self.wait)
                        self._log(r)
                        r.raise_for_status()
                        return r.json()
                    elif kind == 'put':
                        r = requests.put(u, params=args, allow_redirects=False, timeout=self.wait)
                        self._log(r)
                        r.raise_for_status()
                        r = requests.put(r.headers['location'], headers={'content-type': 'application/octet-stream'}, data=data)
                        self._log(r)
                        r.raise_for_status()
                        return True
                    else:
                        r = requests.get(u, params=args, stream=True, timeout=self.wait)
                        self._log(r)
                        r.raise_for_status()
                        for c in r.iter_content(16 * 1024):
                            data.write(c)
                        return True
                except requests.exceptions.HTTPError as e:
                    try:
                        if e.response.json()['RemoteException']['exception'] == 'StandbyException':
                            continue
                        raise WebHDFSError(e.response.json())
                    except ValueError:
                        raise WebHDFSError('%s: %s' % (e.response.reason, path))
                except requests.exceptions.ConnectionError:
                    continue
                except requests.exceptions.Timeout:
                    continue
            else:
                raise WebHDFSConnectionError('cannot connect to any webhdfs endpoint')
        finally:
            self.urls = self.urls[indx:] + self.urls[:indx]

    def _log(self, rsp):
        LOG.debug('url:  %s', rsp.url)
        LOG.debug('code: %d %s', rsp.status_code, rsp.reason)

        w = reduce(lambda x, y: max(x, len(y)), rsp.headers.keys(), 0)
        for k, v in sorted(rsp.headers.iteritems()):
            LOG.debug('  %%-%ds : %%s' % w, k, v)

    def _fix(self, path):
        rval = []

        for part in path.split('/'):
            if not part or part == '.':
                continue
            if rval and part == '..':
                rval.pop()
            else:
                rval.append(part)

        return '/'+'/'.join(rval)

    def stat(self, path, catch=False):
        try:
            r = self._req('GETFILESTATUS', path)
            return WebHDFSObject(path, r['FileStatus'])
        except WebHDFSFileNotFoundError as e:
            if not catch:
                raise e

        return False

    def ls(self, path, recurse=False, request=False):
        p = self._fix(path)
        r = self._req('LISTSTATUS', p)
        for i in r['FileStatuses']['FileStatus']:
            o = WebHDFSObject(p, i)
            if not callable(request) or request(o):
                yield o
                if recurse and o.is_dir():
                    for o in self.ls('%s/%s' % (p, o.name), recurse):
                        yield o

    def glob(self, path):
        l = ['']
        p = self._fix(path)
        c = p.lstrip('/').split('/')
        for i, n in enumerate(c):
            d = []
            for t in l:
                r = self._req('LISTSTATUS', t)
                for f in r['FileStatuses']['FileStatus']:
                    if fnmatch.fnmatch(f['pathSuffix'], n):
                        if i == len(c) - 1:
                            d.append(WebHDFSObject(t, f))
                        elif f['type'] == 'DIRECTORY':
                            d.append('%s/%s' % (t, f['pathSuffix']))
            l = d

        if not l:
            raise WebHDFSFileNotFoundError('%s: no matching file or directory' % p)

        return l

    def du(self, path, real=False):
        p = self._fix(path)
        r = self._req('GETCONTENTSUMMARY', p)
        return r['ContentSummary']['length'] if not real else r['ContentSummary']['spaceConsumed']

    def mkdir(self, path):
        p = self._fix(path)
        r = self._req('MKDIRS', p, 'put')
        return r['boolean']

    def mv(self, path, dest):
        p = self._fix(path)
        d = self._fix(dest)
        r = self._req('RENAME', p, 'put', destination=d)
        return r['boolean']

    def rm(self, path):
        p = self._fix(path)
        r = self._req('DELETE', p, 'delete')
        return r['boolean']

    def repl(self, path, num):
        p = self._fix(path)
        r = self._req('SETREPLICATION', p, 'put', replication=num)
        return r['boolean']

    def get(self, path, data=None):
        rval = True
        if not data:
            rval = False
            data = tempfile.TemporaryFile()

        p = self._fix(path)
        self._req('OPEN', p, 'get', data=data)

        data.flush()
        if os.fstat(data.fileno()).st_size != self.stat(p).size:
            raise WebHDFSIncompleteTransferError('%s: download incomplete' % p)

        if not rval:
            data.seek(0)
            rval = data.read()

        data.close()
        return rval

    def put(self, path, data):
        if isinstance(data, str):
            temp = tempfile.TemporaryFile()
            temp.write(data)
            temp.flush()
            temp.seek(0)

            LOG.debug('%s: saved %d bytes to temp file', temp.name, len(data))
            data = temp

        p = self._fix(path)
        self._req('CREATE', p, 'put', data=data)
        if os.fstat(data.fileno()).st_size != self.stat(p).size:
            raise WebHDFSIncompleteTransferError('%s: upload incomplete' % data.name)

        data.close()
        return True

    def content_summary(self, path):
        p = self._fix(path)
        resp = self._req('GETCONTENTSUMMARY', p, 'get')
        if 'ContentSummary' in resp:
            return resp['ContentSummary']
        else:
            return None
